cmd/lakectl/cmd/fs_upload.go

```diff
@@ -33,7 +33,12 @@ var fsUploadCmd = &cobra.Command{
 		ctx, stop := signal.NotifyContext(ctx, os.Interrupt, os.Kill)
 		defer stop()
 
-		if !recursive { // Assume source is a single file
+		stat, err := os.Stat(source)
+		if err != nil {
+			Die("failed to stat source", 1)
+		}
+
+		if !recursive || !stat.IsDir() { // Ignore recursive if source is a file and not a directory
```

Comment:
I don't understand the code, but I didn't go through all the nooks and crannies.  Suppose `!recursive && stat.IsDir()`.  Then we don't perform this single-object upload, and instead go to l. 53 which says "try recursive upload".  Do we need to fix the comment, or will this really recursively upload directories just because they're directories?

Reply:
If `!recursive && stat.IsDir()` we will get the error:
```
read <path>: is a directory
```

We can explicitly check this matrix and fail if this makes it more clear.